File: airbyte-commons-worker/src/main/java/io/airbyte/workers/helper/AirbyteMessageDataExtractor.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.workers.helper;

import io.airbyte.protocol.models.AirbyteMessage;
import io.airbyte.protocol.models.AirbyteTraceMessage;
import io.airbyte.protocol.models.StreamDescriptor;
import jakarta.inject.Singleton;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Utility class that provides helper functions for extracting data from a {@link AirbyteMessage}.
 */
@Singleton
public class AirbyteMessageDataExtractor {

  private static final Logger LOGGER = LoggerFactory.getLogger(AirbyteMessageDataExtractor.class);

  /**
   * Extracts the stream descriptor from the provided {@link AirbyteMessage}, returning the provided
   * {@code defaultValue} value if the {@link AirbyteMessage} does not include a stream descriptor.
   *
   * @param airbyteMessage The {@link AirbyteMessage} that may contain stream information.
   * @param defaultValue The default value to return if the provided {@link AirbyteMessage} does not
   *        contain stream information.
   * @return The {@link StreamDescriptor} extracted from the provided {@link AirbyteMessage} or the
   *         provided {@code defaultValue} if the message does not contain stream information.
   */
  public StreamDescriptor extractStreamDescriptor(final AirbyteMessage airbyteMessage, final StreamDescriptor defaultValue) {
    final StreamDescriptor extractedStreamDescriptor = getStreamFromMessage(airbyteMessage);
    return extractedStreamDescriptor != null ? extractedStreamDescriptor : defaultValue;
  }

  public StreamDescriptor getStreamFromMessage(final AirbyteMessage airbyteMessage) {
    switch (airbyteMessage.getType()) {
      case RECORD:
        return new StreamDescriptor().withName(airbyteMessage.getRecord().getStream()).withNamespace(airbyteMessage.getRecord().getNamespace());
      case STATE:
        LOGGER.debug("Extracting stream from state message: {}", airbyteMessage.getState().getStream());
        return airbyteMessage.getState().getStream() != null ? airbyteMessage.getState().getStream().getStreamDescriptor() : null;
      case TRACE:
        return getStreamFromTrace(airbyteMessage.getTrace());
      default:
        return null;
    }
  }

  private StreamDescriptor getStreamFromTrace(final AirbyteTraceMessage airbyteTraceMessage) {
    switch (airbyteTraceMessage.getType()) {
      case STREAM_STATUS:
        LOGGER.debug("Extracting stream from stream status trace message: {}", airbyteTraceMessage.getStreamStatus());
        return airbyteTraceMessage.getStreamStatus().getStreamDescriptor();
      case ERROR:
        LOGGER.debug("Extracting stream from error trace message: {}", airbyteTraceMessage.getError());
        return airbyteTraceMessage.getError().getStreamDescriptor();
      case ESTIMATE:
        LOGGER.debug("Extracting stream from estimate trace message: {}", airbyteTraceMessage.getEstimate());
        return new StreamDescriptor().withName(airbyteTraceMessage.getEstimate().getName())
            .withNamespace(airbyteTraceMessage.getEstimate().getNamespace());
      default:
        return null;
    }
  }

}

File: airbyte-container-orchestrator/src/main/kotlin/io/airbyte/container/orchestrator/worker/util/AirbyteMessageDataExtractor.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.container.orchestrator.worker.util

import io.airbyte.protocol.models.v0.AirbyteMessage
import io.airbyte.protocol.models.v0.AirbyteTraceMessage
import io.airbyte.protocol.models.v0.StreamDescriptor
import io.github.oshai.kotlinlogging.KotlinLogging
import jakarta.inject.Singleton

private val logger = KotlinLogging.logger { }

/**
 * Utility class that provides helper functions for extracting data from a [AirbyteMessage].
 */
@Singleton
class AirbyteMessageDataExtractor {
  fun getStreamFromMessage(airbyteMessage: AirbyteMessage): StreamDescriptor? =
    when (airbyteMessage.type) {
      AirbyteMessage.Type.RECORD ->
        StreamDescriptor()
          .withName(airbyteMessage.record.stream)
          .withNamespace(airbyteMessage.record.namespace)

      AirbyteMessage.Type.STATE -> {
        logger.debug { "Extracting stream from state message: ${airbyteMessage.state.stream}" }
        if (airbyteMessage.state.stream != null) airbyteMessage.state.stream.streamDescriptor else null
      }

      AirbyteMessage.Type.TRACE -> getStreamFromTrace(airbyteMessage.trace)
      else -> null
    }
}

private fun getStreamFromTrace(airbyteTraceMessage: AirbyteTraceMessage): StreamDescriptor? =
  when (airbyteTraceMessage.type) {
    AirbyteTraceMessage.Type.STREAM_STATUS -> {
      logger.debug { "Extracting stream from stream status trace message: ${airbyteTraceMessage.streamStatus}" }
      airbyteTraceMessage.streamStatus.streamDescriptor
    }
    AirbyteTraceMessage.Type.ERROR -> {
      logger.debug { "Extracting stream from error trace message: ${airbyteTraceMessage.error}" }
      airbyteTraceMessage.getError().streamDescriptor
    }

    AirbyteTraceMessage.Type.ESTIMATE -> {
      logger.debug { "Extracting stream from estimate trace message: ${airbyteTraceMessage.estimate}" }
      StreamDescriptor()
        .withName(airbyteTraceMessage.estimate.name)
        .withNamespace(airbyteTraceMessage.estimate.namespace)
    }

    else -> null
  }
